File: leetcode/src/main/java/leetcode/array/t2pointer/MaximumPrimeDifference.java
package leetcode.array.t2pointer;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * https://leetcode.com/problems/maximum-prime-difference/
 *
 * Cho mảng số nguyên. Giả sử arr[i], arr[j] là 2 số nguyên tố. Tìm giá trị lớn nhất của |arr[i] - arr[j]|.
 *
 * Ví dụ: nums = [4,2,9,5,3] -> Kết quả là 3
 *
 * Vì nums[1], nums[3], và nums[4] là số nguyên tố. Nên |4 - 1| = 3 là GTLN.
 */
public class MaximumPrimeDifference {

    public boolean primeCheck(int n) {
        if (n < 2) return false;
        for (int i = 2; i * i <= n; ++i) {
            if (n % i == 0) return false;
        }
        return true;
    }

    public int maximumPrimeDifference(int[] nums) {
        int n = nums.length;
        int l = 0, r = n - 1;
        while (l <= r) {
            boolean leftCheck = primeCheck(nums[l]);
            // Tìm được cả 2 vị trí cùng là SNT
            if (leftCheck && primeCheck(nums[r])) {
                return r - l;
            } else if (!leftCheck) { // Số bên trái không phải SNT -> tăng l
                l++;
            } else { // Số bên phải không phải STN -> giảm r
                r--;
            }
        }
        return 0;
    }

    @Test
    public void test() {
        assertEquals(3, maximumPrimeDifference(new int[]{4, 2, 9, 5, 3}));
        assertEquals(0, maximumPrimeDifference(new int[]{4, 8, 2, 8}));
    }

//    public int maximumPrimeDifference_OK(int[] nums) {
//        int n = nums.length;
//        boolean[] mark = new boolean[n];
//        for (int i = 0; i < n; i++) {
//            if (primeCheck(nums[i])) {
//                mark[i] = true;
//            }
//        }
//        int l = 0, r = n - 1;
//        while (l <= r) {
//            if (mark[l] && mark[r]) {
//                return Math.abs(l - r);
//            } else if (mark[l] == false) {
//                l++;
//            } else {
//                r--;
//            }
//        }
//        return 0;
//    }

}

File: py-algorithm/array/t2pointer/maximum_prime_difference.py
import math
import unittest

from math import gcd
from fractions import Fraction
from typing import List
from queue import PriorityQueue
from collections import Counter

"""
    https://leetcode.com/problems/maximum-prime-difference/
    Cho mảng số nguyên. Giả sử arr[i], arr[j] là 2 số nguyên tố. Tìm giá trị lớn nhất của |arr[i] - arr[j]|.
 
    Ví dụ: nums = [4,2,9,5,3] -> Kết quả là 3
 
    Vì nums[1], nums[3], và nums[4] là số nguyên tố. Nên |4 - 1| = 3 là GTLN.
"""


class Solution(unittest.TestCase):

    def prime_check(self, n: int) -> bool:
        if n < 2: return False
        _sqrt = int(math.sqrt(n))
        for i in range(2, _sqrt + 1):
            if n % i == 0: return False

        return True

    def maximumPrimeDifference(self, nums: List[int]) -> int:
        l = 0
        r = len(nums) - 1
        while l <= r:
            left_check = self.prime_check(nums[l])
            if left_check and self.prime_check(nums[r]):
                return r - l
            elif not left_check:
                l += 1
            else:
                r -= 1

        return 0

    def test(self):
        self.assertEqual(3, self.maximumPrimeDifference([4, 2, 9, 5, 3]))
        self.assertEqual(0, self.maximumPrimeDifference([4, 8, 2, 8]))


if __name__ == '__main__':
    unittest.main()
